equinox/nn/attention.py

```diff
@@ -0,0 +1,237 @@
+    $W^K_i \in \mathbb{R}^{d_\text{key} \times d_\text{qk}}$,
+    $W^V_i \in \mathbb{R}^{d_\text{value} \times d_\text{vo}}$,
+    $W^O_i \in \mathbb{R}^{d_\text{vo} \times d_\text{output}}$,
+    with $i \in \{1, \ldots, h\}$, where $h$ is the number of heads, and $d_\text{qk}$,
+    $d_\text{vo}$, $d_\text{output}$ are hyperparameters.
+
+    - $\text{Attention}$ is defined as
+      $\text{Attention}(\widetilde{Q}, \widetilde{K}, \widetilde{V})
+       = \text{softmax}(\frac{\widetilde{Q}\widetilde{K}^\intercal}
+                             {\sqrt{d_\text{qk}}})\widetilde{V}$.
+
+    ??? cite
+
+        [Attention is All You Need](https://arxiv.org/abs/1706.03762)
+
+        ```bibtex
+        @inproceedings{vaswani2017attention,
+            author={Vaswani, Ashish and Shazeer, Noam and Parmar, Niki and
+                    Uszkoreit, Jakob and Jones, Llion and Gomez, Aidan N and
+                    Kaiser, {\L}ukasz and Polosukhin, Illia},
+            booktitle={Advances in Neural Information Processing Systems},
+            publisher={Curran Associates, Inc.},
+            title={Attention is All You Need},
+            volume={30},
+            year={2017}
+        }
+        ```
+
+    !!! faq "FAQ"
+
+        Different software libraries often implement multihead attention in slightly
+        different ways. Some of them will or won't add on biases by default. Most of
+        them will fix the values of $d_\text{qk}, d_\text{vo}, d_\text{output}$ in
+        terms of $d_\text{query}$ or $d_\text{key}$ or $d_\text{value}$. Equinox
+        chooses to expose all of these as options.
+
+        Relative to the original
+        [Attention is All You Need](https://arxiv.org/abs/1706.03762) paper: our
+        $d_\text{qk}$ is their "$d_k$". Our $d_\text{vo}$ is their "$d_\text{v}$". They
+        fix $d_\text{query} = d_\text{key} = d_\text{value} = d_\text{output}$ and
+        refer to it as "$d_\text{model}$".
+    """
+
+    query_proj: Linear
+    key_proj: Linear
+    value_proj: Linear
+    output_proj: Linear
+    dropout: Dropout
+
+    num_heads: int = static_field()
+    query_size: int = static_field()
+    key_size: int = static_field()
+    value_size: int = static_field()
+    output_size: int = static_field()
+    qk_size: int = static_field()
+    vo_size: int = static_field()
+    use_query_bias: bool = static_field()
+    use_key_bias: bool = static_field()
+    use_value_bias: bool = static_field()
+    use_output_bias: bool = static_field()
+
+    def __init__(
+        self,
+        num_heads: int,
+        query_size: int,
+        key_size: Optional[int] = None,
+        value_size: Optional[int] = None,
+        output_size: Optional[int] = None,
+        qk_size: Optional[int] = None,
+        vo_size: Optional[int] = None,
+        use_query_bias: bool = False,
+        use_key_bias: bool = False,
+        use_value_bias: bool = False,
+        use_output_bias: bool = False,
```

Comment:
oh, this should default to True, very rarely is this turned off

Reply:
Will do. Do you mean all four, or just `use_output_bias`?

If you mean just `use_output_bias`, would it be reasonable to change the default on the other three just for consistency?

Reply:
just use_output_bias, the rest is fine!

Reply:
Yeah, the original paper has no biases at all. But if you reckon on having a bias being a more useful/expressive default then I'm happy to keep it this way. (It's already merged now anyway haha!)

Reply:
@patrick-kidger I was wrong, it is better to have it turned off https://storage.googleapis.com/pathways-language-model/PaLM-paper.pdf :pray: sorry about that!

Reply:
Oh no! Thanks for reporting back on that. I'll try and change this in the next breaking release.